### PR TITLE
Authorization for Subscriptions

### PR DESCRIPTION
**Pull request by darrellwarde**
# Description

Adds a new directive, `@subscriptionsAuthorization` which can be used to apply authorization `filter` rules to Subscriptions events.

## Complexity

Complexity: Medium

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-3565.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 63 | 58 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 112 | 123 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 83 | 84 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 5121 | 4522 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 3946 | 4055 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 50 | 51 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 61 | 67 | 2174 |
|  aggregations.DeeplyNestedCount | 13070331 | 13070331 | 6314 | 6103 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 227 | 185 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 80 | 104 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 243 | 290 | 3003 |
|  connections.Connection | 14082 | 14082 | 64 | 83 | 2174 |
|  connections.NestedConnection | 45466 | 45468 | 118 | 166 | 4516 |
|  connections.ConnectionWithSort | 3284 | 3284 | 74 | 91 | 1040 |
|  connections.ConnectionWithSortAndCypher | 3284 | 3284 | 274 | 470 | 1040 |
|  create.SimpleMutation | 7 | 7 | 72 | 101 | 1 |
|  cypher-directive.TopLevelCypherDirective | 122100 | 122100 | 731 | 860 | 12241 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 153619 | 153617 | 243 | 241 | 12241 |
|  delete.SimpleDelete | 19401 | 19401 | 324 | 497 | 1040 |
|  delete.NestedDeleteInUpdate | 23546 | 23546 | 208 | 229 | 2040 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 59 | 73 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 126 | 175 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 55 | 55 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 97 | 88 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 33 | 44 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 61 | 105 | 2174 |
|  query.QueryWhere | 9695 | 9725 | 48 | 74 | 2161 |
|  query.SimpleQueryWithNestedWhere | 9873 | 9903 | 74 | 95 | 2161 |
|  query.Nested | 10096041 | 10096041 | 12089 | 12922 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 11739 | 11968 | 2004000 |
|  query.OrFilterOnRelationships | 43936 | 43986 | 223 | 247 | 2147 |
|  query.OrFilterOnRelationshipsAndNested | 37876 | 37689 | 294 | 325 | 2147 |
|  query.QueryWithNestedIn | 13887 | 13234 | 77 | 89 | 1379 |
|  query.NestedConnectionWhere | 9834 | 9834 | 73 | 84 | 2174 |
|  query.DeeplyNestedConnectionWhere | 10019 | 9943 | 132 | 112 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19326 | 19188 | 193 | 260 | 1658 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 226 | 297 | 1134 |
|  query.Fulltext | 80 | 80 | 36 | 43 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 54 | 64 | 84 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 80 | 100 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1442 | 1469 | 335 | 420 | 1040 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 72 | 76 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 119 | 169 | 4516 |
|  sorting.SortWithTopLevelCypher | 3121 | 3121 | 199 | 291 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 91 | 110 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 87 | 78 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 368 | 365 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 379 | 350 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 226 | 213 | 2002 |

</details>
        
Old Schema Generation: 1:00.916 (m:ss.mmm)
Schema Generation: 1:03.096 (m:ss.mmm)
Old Subgraph Schema Generation: 1:13.499 (m:ss.mmm)
Subgraph Schema Generation: 1:17.231 (m:ss.mmm)

File: packages/graphql/src/schema/resolvers/subscriptions/where/filters/filter-by-authorization-rules.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                        nodes,
+                        relationshipFields,
+                        context,
+                    });
+                });
+            }
+
+            if (!comparisonResultsAggregationFn(comparisonResults)) {
+                return false;
+            }
```
Comment (mjfwebb) [inline packages/graphql/src/schema/resolvers/subscriptions/where/filters/filter-by-authorization-rules.ts:83]: This code block is missing a lot of inferences from Typescript, resulting in many `any`, which makes `comparisonResults` here really hard to reason about. I can't tell what type it is, or what type comparisonResultsAggregationFn is.

File: packages/graphql/src/schema/resolvers/subscriptions/where/filters/filter-by-authorization-rules.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            if (!nodes || !relationshipFields || !isRelationshipSubscriptionsEvent(event)) {
+                return false;
+            }
+
+            const receivedEventRelationshipType = event.relationshipName;
+            const relationships = node.relationFields.filter((f) => f.type === receivedEventRelationshipType);
+            if (!relationships.length) {
+                return false;
+            }
+            const receivedEventRelationship = relationships[0] as RelationField; // ONE relationship only possible
```
Comment (mjfwebb) [inline packages/graphql/src/schema/resolvers/subscriptions/where/filters/filter-by-authorization-rules.ts:129]:
```suggestion
    const receivedEventRelationship = relationships[0]; // ONE relationship only possible
    if (!receivedEventRelationship) {
        return false;
    }
```

With this we can allow Typescript to narrow the type to `RelationField`

File: packages/graphql/src/schema/resolvers/subscriptions/where/utils/filter-relationship-key.ts [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        edgeProperty &&
+        !filterRelationshipEdgeProperty({
+            relationshipFields,
+            relationshipPropertiesInterfaceName,
+            edgeProperty,
+            receivedEventProperties,
+        })
+    ) {
+        return false;
+    }
```
Comment (mjfwebb) [inline packages/graphql/src/schema/resolvers/subscriptions/where/utils/filter-relationship-key.ts:71]: If we assign the result of the `filterRelationshipEdgeProperty()` call to a variable with a meaningful name it becomes a lot easier to read. Not sure if `hasRelationshipEdgeProperty` is correct here, but here's an example of what I mean:

```suggestion
    const hasRelationshipEdgeProperty =
        edgeProperty &&
        !filterRelationshipEdgeProperty({
            relationshipFields,
            relationshipPropertiesInterfaceName,
            edgeProperty,
            receivedEventProperties,
        });
    if (hasRelationshipEdgeProperty) {
        return false;
    }
```

Comment (mjfwebb) [reply packages/graphql/src/schema/resolvers/subscriptions/where/utils/filter-relationship-key.ts:71]: There are several instances of similar `if` cases where it isn't particularly clear what is actually being checked.

**Review by mjfwebb — COMMENTED** [review]
Hopefully some useful comments. My main concern is that reading the code is a little difficult.

File: packages/graphql/src/schema-model/annotation/SubscriptionsAuthorizationAnnotation.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    relationship?: GraphQLWhereArg;
+};
+
+export const SubscriptionsAuthorizationFilterEventRule: ReadonlyArray<SubscriptionsAuthorizationFilterEvent> = [
+    "CREATE",
+    "UPDATE",
+    "DELETE",
+    "CREATE_RELATIONSHIP",
+    "DELETE_RELATIONSHIP",
+];
```
[the remark this answers is not in the document]
Comment (darrellwarde) [reply packages/graphql/src/schema-model/annotation/SubscriptionsAuthorizationAnnotation.ts]: Love this, made this change to both annotations that use it!

File: packages/graphql/src/schema/resolvers/subscriptions/where/filters/filter-by-authorization-rules.ts (hunk 2 above)
Comment (darrellwarde) [reply packages/graphql/src/schema/resolvers/subscriptions/where/filters/filter-by-authorization-rules.ts]: Nice!

File: packages/graphql/src/schema/resolvers/subscriptions/where/filters/filter-by-authorization-rules.ts (hunk 1 above)
Comment (darrellwarde) [reply packages/graphql/src/schema/resolvers/subscriptions/where/filters/filter-by-authorization-rules.ts:83]: Honestly, I was really struggling with this code when I wrote it. It inherited from the existing Subscriptions events filtering, but then added _more_. The typings were crazy as they were, so adding more functionality was a nightmare. Happy to accept a hand through this if you want to have a crack at it!

**Comment by neo4j-team-graphql** [comment]
Looks like you've updated the documentation!

Check out your changes at https://neo4j-graphql-docs-3565.surge.sh

**Review by mjfwebb — APPROVED** [review]
Thanks for the hard work!

**Pull request merged**